### ISSUE TITLE
Keras Mixed Precision

### ISSUE THREAD

**Issue by agg-aditi** [tensorflow/tensorflow#40573]
**System information**
- Linux Ubuntu 18
- TensorFlow installed from (source or binary): Source
- TensorFlow version: 1.14
- Python version: 3.6
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 10.0
- GPU model and memory: 2080 Ti

**Describe the problem**
I am trying to run a simple cnn on Tensor Cores of 2080 Ti. Using the following
from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)
Running into this error
![Screenshot from 2020-06-16 16-03-04](https://user-images.githubusercontent.com/66824777/85017922-e93dd180-b189-11ea-90bd-759f2021b451.png)

**Provide the exact sequence of commands / steps that you executed before running into the problem**

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@agg-aditi 
Please share the code for which this issue is faced.

**Comment by agg-aditi** [comment]
i followed these codes
https://www.tensorflow.org/guide/mixed_precision
https://developer.nvidia.com/automatic-mixed-precision

**Comment by Saduf2019** [comment]
@agg-aditi
Please share simple stand alone code for us tpo replicate or if possible share a colab gist for us to analyse.

**Comment by agg-aditi** [comment]
[keras_mp.txt](https://github.com/tensorflow/tensorflow/files/4804368/keras_mp.txt)

**Comment by agg-aditi** [comment]
import tensorflow as tf

from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.mixed_precision import experimental as mixed_precision


policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)




print('Compute dtype: %s' % policy.compute_dtype)
print('Variable dtype: %s' % policy.variable_dtype)




inputs = keras.Input(shape=(784,), name='digits')
if tf.config.list_physical_devices('GPU'):
  print('The model will run with 4096 units on a GPU')
  num_units = 4096
else:
  # Use fewer units on CPUs so the model finishes in a reasonable amount of time
  print('The model will run with 64 units on a CPU')
  num_units = 64
dense1 = layers.Dense(num_units, activation='relu', name='dense_1')
x = dense1(inputs)
dense2 = layers.Dense(num_units, activation='relu', name='dense_2')
x = dense2(x)



print('x.dtype: %s' % x.dtype.name)

print('dense1.kernel.dtype: %s' % dense1.kernel.dtype.name)






outputs = layers.Dense(10, activation='softmax', name='predictions')(x)
print('Outputs dtype: %s' % outputs.dtype.name)

x = layers.Dense(10, name='dense_logits')(x)
outputs = layers.Activation('softmax', dtype='float32', name='predictions')(x)
print('Outputs dtype: %s' % outputs.dtype.name)



outputs = layers.Activation('linear', dtype='float32')(outputs)




model = keras.Model(inputs=inputs, outputs=outputs)
model.compile(loss='sparse_categorical_crossentropy',
              optimizer=keras.optimizers.RMSprop(),
              metrics=['accuracy'])

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train = x_train.reshape(60000, 784).astype('float32') / 255
x_test = x_test.reshape(10000, 784).astype('float32') / 255



initial_weights = model.get_weights()





history = model.fit(x_train, y_train,
                    batch_size=8192,
                    epochs=5,
                    validation_split=0.2)
test_scores = model.evaluate(x_test, y_test, verbose=2)
print('Test loss:', test_scores[0])
print('Test accuracy:', test_scores[1])



x = tf.constant(256, dtype='float16')
(x ** 2).numpy() 


x = tf.constant(1e-5, dtype='float16')
(x ** 2).numpy() 

loss_scale = policy.loss_scale
print('Loss scale: %s' % loss_scale)




new_policy = mixed_precision.Policy('mixed_float16', loss_scale=1024)
print(new_policy.loss_scale)





optimizer = keras.optimizers.RMSprop()
optimizer = mixed_precision.LossScaleOptimizer(optimizer, loss_scale='dynamic')


loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
train_dataset = (tf.data.Dataset.from_tensor_slices((x_train, y_train))
                 .shuffle(10000).batch(8192))
test_dataset = tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(8192)



@tf.function
def train_step(x, y):
  with tf.GradientTape() as tape:
    predictions = model(x)
    loss = loss_object(y, predictions)
    scaled_loss = optimizer.get_scaled_loss(loss)
  scaled_gradients = tape.gradient(scaled_loss, model.trainable_variables)
  gradients = optimizer.get_unscaled_gradients(scaled_gradients)
  optimizer.apply_gradients(zip(gradients, model.trainable_variables))
  return loss



@tf.function
def test_step(x):
  return model(x, training=False)
a


model.set_weights(initial_weights)



for epoch in range(5):
  epoch_loss_avg = tf.keras.metrics.Mean()
  test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
      name='test_accuracy')
  for x, y in train_dataset:
    loss = train_step(x, y)
    epoch_loss_avg(loss)
  for x, y in test_dataset:
    predictions = test_step(x)
    test_accuracy.update_state(y, predictions)
  print('Epoch {}: loss={}, test accuracy={}'.format(epoch, epoch_loss_avg.result(), test_accuracy.result()))

**Comment by agg-aditi** [comment]
import tensorflow as tf

from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.mixed_precision import experimental as mixed_precision



policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)



print('Compute dtype: %s' % policy.compute_dtype)
print('Variable dtype: %s' % policy.variable_dtype)




inputs = keras.Input(shape=(784,), name='digits')
if tf.config.list_physical_devices('GPU'):
  print('The model will run with 4096 units on a GPU')
  num_units = 4096
else:

  print('The model will run with 64 units on a CPU')
  num_units = 64
dense1 = layers.Dense(num_units, activation='relu', name='dense_1')
x = dense1(inputs)
dense2 = layers.Dense(num_units, activation='relu', name='dense_2')
x = dense2(x)


print('x.dtype: %s' % x.dtype.name)

print('dense1.kernel.dtype: %s' % dense1.kernel.dtype.name)

outputs = layers.Dense(10, activation='softmax', name='predictions')(x)
print('Outputs dtype: %s' % outputs.dtype.name)



x = layers.Dense(10, name='dense_logits')(x)
outputs = layers.Activation('softmax', dtype='float32', name='predictions')(x)
print('Outputs dtype: %s' % outputs.dtype.name)

outputs = layers.Activation('linear', dtype='float32')(outputs)

model = keras.Model(inputs=inputs, outputs=outputs)
model.compile(loss='sparse_categorical_crossentropy',
              optimizer=keras.optimizers.RMSprop(),
              metrics=['accuracy'])

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train = x_train.reshape(60000, 784).astype('float32') / 255
x_test = x_test.reshape(10000, 784).astype('float32') / 255



initial_weights = model.get_weights()



history = model.fit(x_train, y_train,
                    batch_size=8192,
                    epochs=5,
                    validation_split=0.2)
test_scores = model.evaluate(x_test, y_test, verbose=2)
print('Test loss:', test_scores[0])
print('Test accuracy:', test_scores[1])





x = tf.constant(256, dtype='float16')
(x ** 2).numpy()  


x = tf.constant(1e-5, dtype='float16')
(x ** 2).numpy()  



loss_scale = policy.loss_scale
print('Loss scale: %s' % loss_scale)




new_policy = mixed_precision.Policy('mixed_float16', loss_scale=1024)
print(new_policy.loss_scale)



optimizer = keras.optimizers.RMSprop()
optimizer = mixed_precision.LossScaleOptimizer(optimizer, loss_scale='dynamic')



loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
train_dataset = (tf.data.Dataset.from_tensor_slices((x_train, y_train))
                 .shuffle(10000).batch(8192))
test_dataset = tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(8192)



@tf.function
def train_step(x, y):
  with tf.GradientTape() as tape:
    predictions = model(x)
    loss = loss_object(y, predictions)
    scaled_loss = optimizer.get_scaled_loss(loss)
  scaled_gradients = tape.gradient(scaled_loss, model.trainable_variables)
  gradients = optimizer.get_unscaled_gradients(scaled_gradients)
  optimizer.apply_gradients(zip(gradients, model.trainable_variables))
  return loss


@tf.function
def test_step(x):
  return model(x, training=False)


model.set_weights(initial_weights)


for epoch in range(5):
  epoch_loss_avg = tf.keras.metrics.Mean()
  test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
      name='test_accuracy')
  for x, y in train_dataset:
    loss = train_step(x, y)
    epoch_loss_avg(loss)
  for x, y in test_dataset:
    predictions = test_step(x)
    test_accuracy.update_state(y, predictions)
  print('Epoch {}: loss={}, test accuracy={}'.format(epoch, epoch_loss_avg.result(), test_accuracy.result()))

**Comment by Saduf2019** [comment]
@
Code shared is not indented, i ran the code after fixing few lines the entire code is unindented, please find the [gits here](https://colab.research.google.com/gist/Saduf2019/a6e7b8bdcc0e5dd2e2ddacc6e1b8eb96/untitled238.ipynb)
Please share code (indented) such that we can replicate the issue faced or  please share colab gist with the error faced for us to analyse.

**Comment by agg-aditi** [comment]
> @
> Code shared is not indented, i ran the code after fixing few lines the entire code is unindented, please find the [gits here](https://colab.research.google.com/gist/Saduf2019/a6e7b8bdcc0e5dd2e2ddacc6e1b8eb96/untitled238.ipynb)
> Please share code (indented) such that we can replicate the issue faced or please share colab gist with the error faced for us to analyse.

https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/guide/mixed_precision.ipynb

**Comment by Saduf2019** [comment]
@agg-aditi
By colab gist i mean the code you have used, to reproduce the issue as shared in my gist, please remove indentation errors and share.

**Comment by agg-aditi** [comment]
> @agg-aditi
> By colab gist i mean the code you have used, to reproduce the issue as shared in my gist, please remove indentation errors and share.

https://gist.github.com/agg-aditi/5a3727eee3e052e1275bf95e7284a115
this is the exact code i have used

**Comment by agg-aditi** [comment]
i tried running the this tutorial which itself gave the error

> > @
> > Code shared is not indented, i ran the code after fixing few lines the entire code is unindented, please find the [gits here](https://colab.research.google.com/gist/Saduf2019/a6e7b8bdcc0e5dd2e2ddacc6e1b8eb96/untitled238.ipynb)
> > Please share code (indented) such that we can replicate the issue faced or please share colab gist with the error faced for us to analyse.
> > @Saduf2019
> > https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/guide/mixed_precision.ipynb

**Comment by gowthamkpr** [comment]
@agg-aditi I ran the code but didn't run into any error. Heres my [gist](https://colab.research.google.com/gist/gowthamkpr/c9496350467bb22505443c62136827a5/mixed_precision.ipynb). You have to run it using tensorflow 2.2.0 not using 1.X. Thanks!

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive fr more than 2 weeks. Please add additional comments for us to open this issue again. Thanks!
